src/app.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Window building and app lifecycle.
+
+use std::sync::Arc;
+
+use crate::shell::{init, runloop, Error as PlatformError, WindowBuilder};
+use crate::{Data, LocalizedString, UiMain, UiState, Widget};
+
+/// Handles initial setup of an application, and starts the runloop.
+pub struct AppLauncher<T> {
+    windows: Vec<WindowDesc<T>>,
+}
+
+/// A function that can create a widget.
+///
+/// This type signature is a bit hairy because this has to work with `Command`,
+/// which requires things to be sync + send.
+type WidgetBuilderFn<T> = dyn Fn() -> Box<dyn Widget<T>> + Send + Sync + 'static;
+
+/// A description of a window to be instantiated.
+///
+/// This includes a function that can build the root widget, as well as other
+/// window properties such as the title.
+pub struct WindowDesc<T> {
+    pub(crate) root_builder: Arc<WidgetBuilderFn<T>>,
+    //TODO: more things you can configure on a window, like size and menu
+    pub(crate) title: Option<LocalizedString<T>>,
+}
+
+impl<T: Data + 'static> AppLauncher<T> {
+    /// Create a new `AppLauncher` with the provided window.
+    pub fn with_window(window: WindowDesc<T>) -> Self {
+        AppLauncher {
+            windows: vec![window],
+        }
+    }
+
+    /// Build the windows and start the runloop.
+    ///
+    /// Returns an error if a window cannot be instantiated. This is usually
+    /// a fatal error.
+    pub fn launch(mut self, data: T) -> Result<(), PlatformError> {
+        init();
+
+        //TODO: launch all windows, when multi-win lands
+        let window = self.windows.pop().expect("launch called with no window");
+        let WindowDesc {
+            root_builder,
+            title,
+            ..
+        } = window;
+
+        //TODO: use title when multi-win lands
+        let _title = title.unwrap_or(LocalizedString::new("app-name-exclaim"));
```

Comment:
It looks like this regresses setting a title at all. I'm patient with that landing, but if it's a simple change to keep basic (non-localized) functionality in place, that'd be cool.

Reply:
ah fair, changed so it will use the non-localized `localized_str()` value (e.g. the placeholder if one is set, or the key if it isn't).